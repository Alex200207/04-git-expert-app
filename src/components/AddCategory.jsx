import { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ onNewCategory }) => {
  const [inputValue, setInputValue] = useState();

  const onInputChange = (event) => {
    //usamos el evento para obtener el valor del input
    //los eventos contienen un target que es el elemento que disparo el evento
    //informacion sobre la ubicacion valor entre otras
    setInputValue(event.target.value);
  };

  const onSubmit = (event) => {
    console.log("submit");
    event.preventDefault(); //evita que se recargue la pagina

    if (inputValue.trim().length <= 1) return; //esta funcion se va cumplir si el valor del input es menor o igual a 1
    // setCategories((categorias) => [inputValue, ...categorias]);//agrega el nuevo elemento al arreglo de categorias y preserva las anteriores
    onNewCategory(inputValue.trim()); //
    setInputValue(""); //limpia el input para que este vacio despues de agregar la categoria
  };

  return (
    <form onSubmit={onSubmit} aria-label="form">
      <input
        type="text"
        placeholder="Buscar gifs"
        value={inputValue}
        onChange={onInputChange} //cada vez que cambie el input se dispara el evento
      />
    </form>
  );
};

AddCategory.propTypes = {
  onNewCategory: PropTypes.func.isRequired,//la funcion onNewCategory es requerida
};
