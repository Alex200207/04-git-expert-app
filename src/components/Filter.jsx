import { useState } from "react";
import { CiFilter } from "react-icons/ci";

const Filter = ({ limit, onLimitChange }) => {
  const [isOpen, setIsOpen] = useState(false);

  const handleChange = (event) => {
    onLimitChange(Number(event.target.value));
    setIsOpen(false);
  };

  return (
    <div className="filter-container" style={{ position: "relative" }}>
      <CiFilter
        className="search__filter"
        style={{ cursor: "pointer", fontSize: "24px" }}
        onClick={() => setIsOpen(!isOpen)}
      />

      {isOpen && (
        <select
          id="limit-select"
          value={limit}
          onChange={handleChange}
          style={{
            position: "absolute",
            top: "30px",
            left: "0",
            zIndex: 1000,
            width: "100px",
            backgroundColor: "white",
            border: "1px solid #ccc",
            borderRadius: "4px",
            padding: "5px",
            boxShadow: "0 2px 5px rgba(0, 0, 0, 0.15)",
          }}
        >
          <option value={5} style={{ padding: "5px" }}>
            5
          </option>
          <option value={10} style={{ padding: "5px" }}>
            10
          </option>
          <option value={15} style={{ padding: "5px" }}>
            15
          </option>
          <option value={20} style={{ padding: "5px" }}>
            20
          </option>
        </select>
      )}
    </div>
  );
};

export default Filter;
