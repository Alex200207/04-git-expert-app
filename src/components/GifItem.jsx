import PropTypes from "prop-types";
import { IoCopyOutline } from "react-icons/io5";
import { MdOutlineShare } from "react-icons/md";
import { FaWhatsapp } from "react-icons/fa";

export const GifItem = ({ title, url }) => {
  const handleShare = () => {
    navigator.clipboard.writeText(url).then(() => {
      alert("URL copiada al portapapeles");
    });
  };

  return (
    <div className="card">
      <img src={url} alt={title} />
      <p>{title}</p>
      <div className="btn-container">
        <button onClick={handleShare} className="btn-container__share">
          <IoCopyOutline className="btn-container__share_icon" />
        </button>
        <a
          href={`https://api.whatsapp.com/send?text=${url}`}
          className="btn_container__whatsapp"
        >
          <FaWhatsapp className="btn-container__whatssap_icon" />
          <MdOutlineShare className="btn-container__whatssap_icon" />
        </a>
      </div>
    </div>
  );
};

GifItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};
