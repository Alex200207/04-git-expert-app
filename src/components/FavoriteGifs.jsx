import { GifItem } from "./GifItem";

function FavoriteGifs({ favorites, onToggleFavorite }) {
  return (
    <>
      {favorites.map((gif) => (
        <GifItem
          key={gif.id}
          {...gif}
          onToggleFavorite={onToggleFavorite}
          favorites={favorites}
        />
      ))}
    </>
  );
}

export default FavoriteGifs;
