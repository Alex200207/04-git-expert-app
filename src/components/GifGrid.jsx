import PropTypes from "prop-types";
import { GifItem } from "./GifItem";
import { useFetchGifs } from "../hooks/useFetchGifs";
import { MdDelete } from "react-icons/md";

export const GifGrid = ({ category, onDeleteCategory }) => {
  const { images, isLoading } = useFetchGifs(category);
  //crear un estado local
  // const [images, setImages] = useState([]);
  // const getImage = async () => {
  //   const newImage = await getGifs(category);
  //   setImages(newImage);
  // };

  // //use efect es un hook que nos permite ejecutar codigo de manera condicional
  // useEffect(() => {
  //   getImage();
  //   // eslint-disable-next-line react-hooks/exhaustive-deps
  // }, []); //el segundo argumento es un arreglo de dependencias que se ejecutara cuando la dependencia cambie

  return (
    <>
      <div
        style={{
          display: "flex",
          justifyContent: "space-between",
          alignItems: "center",
        }}
      >
        <h3>{category}</h3>
        <button onClick={() => onDeleteCategory(category)}>
          <MdDelete className="btn-delete" />
        </button>
      </div>
      {isLoading && <p>Cargando..</p>}

      <div className="card-grid">
        {images.map(
          (
            image //destructuramos el objeto
          ) => (
            <GifItem key={image.id} {...image} /> //usamos el spread operator para pasar las propiedades del objeto como props
          )
        )}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
};
