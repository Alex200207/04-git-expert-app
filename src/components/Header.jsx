import React, { useState, useEffect } from "react";
import { Heart, X } from "lucide-react";
import Filter from "../components/Filter";
import { AddCategory } from "./AddCategory";
import "../components/style/Header.css";
import FavoriteGifs from "./FavoriteGifs";

const Header = ({
  onNewCategory,
  onLimitChange,
  limit,
  favorites,
  onToggleFavorite,
}) => {
  const [isAsideOpen, setIsAsideOpen] = useState(false);
  const [asideWidth, setAsideWidth] = useState(600);
  const [isResizing, setIsResizing] = useState(false);
  const [isNavbarVisible, setIsNavbarVisible] = useState(true);
  const [lastScrollTop, setLastScrollTop] = useState(0);

  const toggleAside = () => {
    setIsAsideOpen(!isAsideOpen);
  };

  const handleMouseDown = (e) => {
    e.preventDefault();
    setIsResizing(true);
  };

  const handleMouseMove = (e) => {
    if (isResizing) {
      const newWidth = window.innerWidth - e.clientX;
      setAsideWidth(Math.max(600, newWidth));
    }
  };

  const handleMouseUp = () => {
    setIsResizing(false);
  };

  useEffect(() => {
    if (isResizing) {
      window.addEventListener("mousemove", handleMouseMove);
      window.addEventListener("mouseup", handleMouseUp);
    } else {
      window.removeEventListener("mousemove", handleMouseMove);
      window.removeEventListener("mouseup", handleMouseUp);
    }
    return () => {
      window.removeEventListener("mousemove", handleMouseMove);
      window.removeEventListener("mouseup", handleMouseUp);
    };
  }, [isResizing]);

  useEffect(() => {
    const handleScroll = () => {
      const currentScroll = document.documentElement.scrollTop;
      if (currentScroll > lastScrollTop) {
        setIsNavbarVisible(false); // Ocultar navbar
      } else {
        setIsNavbarVisible(true); // Mostrar navbar
      }
      setLastScrollTop(currentScroll <= 0 ? 0 : currentScroll);
    };

    window.addEventListener("scroll", handleScroll);
    return () => {
      window.removeEventListener("scroll", handleScroll);
    };
  }, [lastScrollTop]);

  return (
    <>
      <nav
        className={`navbar-container ${isNavbarVisible ? "visible" : "hidden"}`}
      >
        <div className="container_main-title">
          <h1 className="main-title">GIFapp</h1>
        </div>

        <AddCategory onNewCategory={onNewCategory} />
        <div className="search">
          <Filter onLimitChange={onLimitChange} limit={limit} />
        </div>
        <NavItem
          icon={<Heart color="#ef4444" fill="#ef4444" />}
          label={<span className="nav-label">Favoritos</span>}
          active={true}
          onClick={toggleAside}
          count={favorites.length}
        />
      </nav>

      <aside
        className={`favorites-aside ${isAsideOpen ? "open" : ""}`}
        style={{ width: `${asideWidth}px` }}
      >
        <div className="aside-header">
          <h2>Mis Favoritos</h2>
          <button className="close-button" onClick={toggleAside}>
            <X size={24} />
          </button>
        </div>
        <div className="aside-content">
          {favorites.length === 0 && (
            <p className="aside-content__p">No tienes favoritos</p>
          )}
          <FavoriteGifs
            onToggleFavorite={onToggleFavorite}
            favorites={favorites}
          />
        </div>
        <div className="resizer" onMouseDown={handleMouseDown}></div>
      </aside>

      {isAsideOpen && <div className="overlay" onClick={toggleAside}></div>}
    </>
  );
};

const NavItem = ({ icon, label, active, onClick, count }) => {
  return (
    <button className={`nav-item ${active ? "active" : ""}`} onClick={onClick}>
      <div className="icon-container">
        {icon}
        {count > 0 && <span className="badge">{count}</span>}
      </div>
      <span>{label}</span>
    </button>
  );
};

export default Header;
