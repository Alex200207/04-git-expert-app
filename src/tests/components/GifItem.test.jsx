import { render, screen } from "@testing-library/react";
import { GifItem } from "../../components/GifItem";

describe("prueba en el componente GiftItem", () => {
  const title = "Un titulo";
  const url = "https://localhost/algo.jpg";
  
  test("debe de hacer match con el snapshot", () => {
    const { container } = render(<GifItem title={title} url={url} />);

    expect(container).toMatchSnapshot();
  });


  test('debe de mostrar la imagen con el URL y el ALT indicado', () => {

    render(<GifItem title={title} url={url} />);
    // screen.debug();
    //debug imprime en consola el arbol del componente
    // expect( screen.getByRole('img').src).toBe(url);
    // expect( screen.getByRole('img').alt).toBe(title);

    //forma correcta

    const {src, alt} = screen.getByRole('img');
    expect(alt).toBe(title);
    expect(src).toBe(url);
  });

  test('debe mostrar el titulo en el componente', () => { 
    render(<GifItem title={title} url={url} />);
    expect(screen.getByText(title)).toBeTruthy();// esperamos q el titulo exista
   })
});
//screen es un objeto que nos permite buscar elementos en el arbol del componente
//getByRole es una funcion que nos permite buscar elementos por su rol
//container es el contenedor que contiene el componente renderizado
//toMatchSnapshot es una funcion que toma una instantanea de la salida del componente
//y la compara con la instantanea guardada
//si las instantaneas no coinciden, la prueba falla

/*
Este código es una prueba unitaria para el componente `GifItem` usando Jest y React Testing Library. 
Primero, define un bloque de pruebas con `describe`, que agrupa las pruebas relacionadas con el componente `GifItem`.
Dentro de este bloque, se definen dos constantes `title` y `url` que se usarán como props para el componente.
Luego, se define una prueba con `test` que verifica si el componente `GifItem` coincide con el snapshot guardado previamente.
La función `render` de React Testing Library se usa para renderizar el componente con las props `title` y `url`.
Finalmente, se usa `expect(container).toMatchSnapshot()` para comparar la salida renderizada con el snapshot almacenado.
*/
