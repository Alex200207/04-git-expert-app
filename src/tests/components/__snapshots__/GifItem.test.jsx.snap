// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prueba en el componente GiftItem debe de hacer match con el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Un titulo"
      src="https://localhost/algo.jpg"
    />
    <p>
      Un titulo
    </p>
    <div
      class="btn-container"
    >
      <button
        class="btn-container__share"
      >
        <svg
          class="btn-container__share_icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="none"
            height="336"
            rx="57"
            ry="57"
            stroke-linejoin="round"
            stroke-width="32"
            width="336"
            x="128"
            y="128"
          />
          <path
            d="m383.5 128 .5-24a56.16 56.16 0 0 0-56-56H112a64.19 64.19 0 0 0-64 64v216a56.16 56.16 0 0 0 56 56h24"
            fill="none"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="32"
          />
        </svg>
      </button>
      <a
        class="btn_container__whatsapp"
        href="https://api.whatsapp.com/send?text=https://localhost/algo.jpg"
      >
        <svg
          class="btn-container__whatssap_icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M380.9 97.1C339 55.1 283.2 32 223.9 32c-122.4 0-222 99.6-222 222 0 39.1 10.2 77.3 29.6 111L0 480l117.7-30.9c32.4 17.7 68.9 27 106.1 27h.1c122.3 0 224.1-99.6 224.1-222 0-59.3-25.2-115-67.1-157zm-157 341.6c-33.2 0-65.7-8.9-94-25.7l-6.7-4-69.8 18.3L72 359.2l-4.4-7c-18.5-29.4-28.2-63.3-28.2-98.2 0-101.7 82.8-184.5 184.6-184.5 49.3 0 95.6 19.2 130.4 54.1 34.8 34.9 56.2 81.2 56.1 130.5 0 101.8-84.9 184.6-186.6 184.6zm101.2-138.2c-5.5-2.8-32.8-16.2-37.9-18-5.1-1.9-8.8-2.8-12.5 2.8-3.7 5.6-14.3 18-17.6 21.8-3.2 3.7-6.5 4.2-12 1.4-32.6-16.3-54-29.1-75.5-66-5.7-9.8 5.7-9.1 16.3-30.3 1.8-3.7.9-6.9-.5-9.7-1.4-2.8-12.5-30.1-17.1-41.2-4.5-10.8-9.1-9.3-12.5-9.5-3.2-.2-6.9-.2-10.6-.2-3.7 0-9.7 1.4-14.8 6.9-5.1 5.6-19.4 19-19.4 46.3 0 27.3 19.9 53.7 22.6 57.4 2.8 3.7 39.1 59.7 94.8 83.8 35.2 15.2 49 16.5 66.6 13.9 10.7-1.6 32.8-13.4 37.4-26.4 4.6-13 4.6-24.1 3.2-26.4-1.3-2.5-5-3.9-10.5-6.6z"
          />
        </svg>
        <svg
          class="btn-container__whatssap_icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81 1.66 0 3-1.34 3-3s-1.34-3-3-3-3 1.34-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9c-1.66 0-3 1.34-3 3s1.34 3 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.16c-.05.21-.08.43-.08.65 0 1.61 1.31 2.92 2.92 2.92s2.92-1.31 2.92-2.92c0-1.61-1.31-2.92-2.92-2.92zM18 4c.55 0 1 .45 1 1s-.45 1-1 1-1-.45-1-1 .45-1 1-1zM6 13c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1zm12 7.02c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1z"
          />
        </svg>
      </a>
    </div>
  </div>
</div>
`;
